### DISCUSSION TITLE
Runtime changes

### DISCUSSION THREAD

**Discussion by MentalGear** [matteobruni/tsparticles#1753] [Q&A] [answered]
I'm wondering if changes during runtime would be possible?
I like to change the link distance to expand over time.

**Comment by matteobruni** [comment, marked as answer]
Runtime changes are too heavy, but I can think an animation for links distance, this could be a nice thing to do

**Reply by MentalGear** [reply]
Thanks for the reply, Matteo! Darn, I thought so.
I was hoping to animate an icon starting out with small particles, and then the particle size, movement as well as the link distance would increase to depict increased activity.

Here's the svg I wanted to use, though I also don't seem to get it to load in the "polygone mask" preset. Is there something I'm missing?
https://svgshare.com/i/Yjt.svg

**Reply by matteobruni** [reply]
Try to copy that locally, it has a good format and it should work. The SVG files for the polygon masks are loaded using AJAX requests and CORS could be the error in your case if you are using an external link

**Reply by MentalGear** [reply]
Thanks that was the problem and sorry for getting slightly off topic!
Btw: Amazing Work Matteo; I understand that for some properties reactivity could be heavy, however I hope you might consider some properties where it makes sense, like linkDistance, particlesSize, particleSpeed?

**Reply by matteobruni** [reply]
Particle size has already animation properties, Speed has gravity or speed affected by size. Links distance instead is static and I think I can work on this for the next release

**Reply by MentalGear** [reply]
Thanks for being so forthcoming. :)
I'm not sure though if we are on the same page: I know that some props are animatable in that they can be set to change automatically between a min and max value via config JSON.

What I mean by "reactive" however is that prop values can be manipulated from outside to achieve a certain effect, for example increasing size, speed and link distance from 0 to 100 progressively and then stoping, without it "animating" back and forth.

**Reply by matteobruni** [reply]
About Size and Opacity animations they have a `count` property (https://particles.js.org/docs/classes/options_classes_particles_size_sizeanimation.sizeanimation.html#count) that specifies how many times the animation will be performed, if `1` is used, it will run once, `0` will run forever. Speed gravity has a `maxSpeed` property or you can use the `size` property to decrease particle speed for smaller particles.

**Reply by MentalGear** [reply]
Interesting. Yet having the possibility to programmatically change properties during runtime would bring even greater flexibility and open up tsparticles to even more use cases.

I'm wondering if there's anything that prevents the properties to be simply exposed through the API?
Would also make the editor much more satisfying to play with by actually showing changes live.

**Reply by matteobruni** [reply]
There's nothing that prevents particles to be changed, you can see every particle property here: https://particles.js.org/docs/classes/core_particle.particle.html

I have no plan about live changes since I tried removing stored values but recalculating them every time worsened a lot performances

**Reply by MentalGear** [reply]
> There's nothing that prevents particles to be changed, you can see every particle property here: https://particles.js.org/docs/classes/core_particle.particle.html

I'm not sure I can follow. How would I change particles programatically during runtime? The only thing I found is `tsParticles.set` method, though not sure how to use it.

**Reply by matteobruni** [reply]
`tsParticles.load` returns a Promise with a `Container` class (https://particles.js.org/docs/classes/core_container.container.html), you can save it in a variable and read `.particles.array` property where all particles are stored, every particle is an instance of that `Particle` class

If you use the library in Vanilla Javascript with the CDN you can access also the `tsParticles.domItem(idx)` where idx is the container number (`0` should be fine, if you have multiple instances you have to choose the right index) and it's the same `Container` class

**Reply by MentalGear** [reply]
I played a bit with the `animation` and `count` option as you suggested. It works fine for size, but when trying to animate opacity from 0 to 1 with a count of 1, opacity animates a few times and doesn't stop at 1. Could this be a bug ?

```JS
tsParticles.load("tsparticles", {"autoPlay":true,"background":{"color":{"value":"#0d47a1"},"image":"","position":"50% 50%","repeat":"no-repeat","size":"cover","opacity":1},"backgroundMask":{"composite":"destination-out","cover":{"color":{"value":"#fff"},"opacity":1},"enable":false},"fullScreen":{"enable":true,"zIndex":1},"detectRetina":true,"fpsLimit":60,"infection":{"cure":false,"delay":0,"enable":false,"infections":0,"stages":[]},"interactivity":{"detectsOn":"canvas","events":{"onClick":{"enable":true,"mode":"push"},"onDiv":{"selectors":"#repulse-div","enable":false,"mode":"repulse","type":"circle"},"onHover":{"enable":true,"mode":"repulse","parallax":{"enable":false,"force":60,"smooth":10}},"resize":true},"modes":{"attract":{"distance":200,"duration":0.4,"easing":"ease-out-quad","factor":1,"maxSpeed":50,"speed":1},"bounce":{"distance":200},"bubble":{"distance":400,"duration":2,"opacity":0.8,"size":40},"connect":{"distance":80,"links":{"opacity":0.5},"radius":60},"grab":{"distance":400,"links":{"blink":false,"consent":false,"opacity":1}},"light":{"area":{"gradient":{"start":{"value":"#ffffff"},"stop":{"value":"#000000"}},"radius":1000},"shadow":{"color":{"value":"#000000"},"length":2000}},"push":{"quantity":4},"remove":{"quantity":2},"repulse":{"distance":200,"duration":0.4,"factor":100,"speed":1,"maxSpeed":50,"easing":"ease-out-quad"},"slow":{"factor":1,"radius":0},"trail":{"delay":1,"pauseOnStop":false,"quantity":1}}},"manualParticles":[],"motion":{"disable":false,"reduce":{"factor":4,"value":true}},"particles":{"bounce":{"horizontal":{"random":{"enable":false,"minimumValue":0.1},"value":1},"vertical":{"random":{"enable":false,"minimumValue":0.1},"value":1}},"collisions":{"bounce":{"horizontal":{"random":{"enable":false,"minimumValue":0.1},"value":1},"vertical":{"random":{"enable":false,"minimumValue":0.1},"value":1}},"enable":false,"mode":"bounce","overlap":{"enable":true,"retries":0}},"color":{"value":"#ffffff","animation":{"h":{"count":0,"enable":false,"offset":0,"speed":1,"sync":true},"s":{"count":0,"enable":false,"offset":0,"speed":1,"sync":true},"l":{"count":0,"enable":false,"offset":0,"speed":1,"sync":true}}},"destroy":{"mode":"none","split":{"count":1,"factor":{"random":{"enable":false,"minimumValue":0},"value":3},"rate":{"random":{"enable":false,"minimumValue":0},"value":{"min":4,"max":9}},"sizeOffset":true}},"life":{"count":0,"delay":{"random":{"enable":false,"minimumValue":0},"value":0,"sync":false},"duration":{"random":{"enable":false,"minimumValue":1},"value":0,"sync":false}},"links":{"blink":false,"color":{"value":"#ffffff"},"consent":false,"distance":150,"enable":true,"frequency":1,"opacity":0.4,"shadow":{"blur":5,"color":{"value":"lime"},"enable":false},"triangles":{"enable":false,"frequency":1},"width":1,"warp":false},"move":{"angle":{"offset":0,"value":90},"attract":{"distance":200,"enable":false,"rotate":{"x":600,"y":1200}},"decay":0,"distance":0,"direction":"none","drift":0,"enable":false,"gravity":{"acceleration":9.81,"enable":false,"inverse":false,"maxSpeed":50},"path":{"clamp":true,"delay":{"random":{"enable":false,"minimumValue":0},"value":0},"enable":false},"outModes":{"default":"out","bottom":"out","left":"out","right":"out","top":"out"},"random":false,"size":false,"speed":1,"straight":true,"trail":{"enable":false,"length":10,"fillColor":{"value":"#000000"}},"vibrate":false,"warp":false},"number":{"density":{"enable":true,"area":800,"factor":1000},"limit":0,"value":80},"opacity":{"random":{"enable":false,"minimumValue":0.1},"value":{"min":0,"max":1},"animation":{"count":1,"enable":true,"speed":1,"sync":false,"destroy":"none","minimumValue":0,"startValue":"min"}},"reduceDuplicates":false,"roll":{"darken":{"enable":false,"value":0},"enable":false,"enlighten":{"enable":false,"value":0},"speed":25},"rotate":{"random":{"enable":false,"minimumValue":0},"value":0,"animation":{"enable":false,"speed":0,"sync":false},"direction":"clockwise","path":false},"shadow":{"blur":0,"color":{"value":"#000000"},"enable":false,"offset":{"x":0,"y":0}},"shape":{"options":{"character":[{"fill":true,"font":"Font Awesome 5 Brands","style":"","value":[""],"weight":"400"},{"fill":true,"font":"Font Awesome 5 Free","style":"","value":[""],"weight":"900"}],"char":[{"fill":true,"font":"Font Awesome 5 Brands","style":"","value":[""],"weight":"400"},{"fill":true,"font":"Font Awesome 5 Free","style":"","value":[""],"weight":"900"}],"polygon":{"sides":5},"star":{"sides":5},"image":{"height":100,"replaceColor":true,"src":"https://particles.js.org/images/github.svg","width":100},"images":{"height":100,"replaceColor":true,"src":"https://particles.js.org/images/github.svg","width":100}},"type":"circle"},"size":{"random":{"enable":false,"minimumValue":0},"value":{"min":1,"max":25},"animation":{"count":1,"enable":true,"speed":5,"sync":true,"destroy":"none","minimumValue":1,"startValue":"min"}},"stroke":{"width":1,"color":{"value":"#ffffff","animation":{"h":{"count":0,"enable":false,"offset":0,"speed":1,"sync":true},"s":{"count":0,"enable":false,"offset":0,"speed":1,"sync":true},"l":{"count":0,"enable":false,"offset":0,"speed":1,"sync":true}}}},"tilt":{"random":{"enable":false,"minimumValue":0},"value":0,"animation":{"enable":false,"speed":0,"sync":false},"direction":"clockwise","enable":false},"twinkle":{"lines":{"enable":false,"frequency":0.05,"opacity":1},"particles":{"enable":false,"frequency":0.05,"opacity":1}},"wobble":{"distance":5,"enable":false,"speed":50}},"pauseOnBlur":true,"pauseOnOutsideViewport":true,"responsive":[],"themes":[]});
```

**Reply by matteobruni** [reply]
Yes it's a bug, I found that now. Thanks! I'm going to release a fix ASAP

**Reply by matteobruni** [reply]
> Let me just say your support and maintenance is truly extraordinary! Let me know if I can tip you a drink or something somewhere - really this should be an integrated functions in the discucions feature by github.

Thank you very much, I have a lot of fun with this project so I spend a lot of time here hoping to not lose interest over time. If you want to donate or something like that I have enabled the sponsor profile some time ago, you can see the button in the project homepage

> If you maybe find the time to integrate animation for link and motion sometime, that would be appreciated by me as well.

I'm preparing 1.31 right now, I'm doing some planned changes for now, hope to bring those features too

**Reply by matteobruni** [reply]
@MentalGear  `1.30.4` is out with the opacity animation fix

**Comment by MentalGear** [comment]
This would allow manipulation during runtime ?

**Reply by matteobruni** [reply]
Yes it's what the library does under the hood

**Reply by MentalGear** [reply]
Thx for the quick answer, I'll try it out and get back to you.
